### PR TITLE
core: preallocate RPC argument

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Preallocates an RPC argument structure when starting a new thread. Since
a thread can only have one active RPC at a time it's enough to allocate
one RPC argument for all RPC during the lifetime of the thread. The
preallocated RPC argument is used internally by thread_rpc_cmd(). This
simplifies all calls to thread_rpc_cmd().

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (QEMU)

This is a cherry-pick of the first commit in https://github.com/OP-TEE/optee_os/pull/319

### CODE REVIEW

File: core/arch/arm/kernel/thread.c [hunk 1]
```diff
@@ -553,6 +555,39 @@ void thread_handle_std_smc(struct thread_smc_args *args)
...
+		thr->rpc_parg = parg;
+	}
+
+	thread_std_smc_handler_ptr(args);
+
+	/*
+	 * Free the RPC argument to avoid having allocated shared memory
+	 * when all threads are done.
+	 */
+	thread_rpc_free_arg(thr->rpc_parg);
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread.c]: Perhaps we should remove this step as it causes the allocation above to be performed on each entry.

Comment (pascal-brand-st-dev) [inline core/arch/arm/kernel/thread.c]: Agree

**Comment by pascal-brand-st-dev** [comment]
Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

File: core/arch/arm/kernel/thread.c (hunk 1 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread.c]: OK, I'll remove it.

**Comment by jenswi-linaro** [comment]
Review tag applied. Removed the rpc_free as agreed.

**Comment by pascal-brand-st** [comment]
Manually merged

**Pull request closed without merging**